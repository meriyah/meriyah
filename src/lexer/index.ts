export { scanSingleToken, nextToken } from './scan';
export { skipHashBang } from './comments';
export { advanceChar, consumeLineFeed, scanNewLine, LexerState, convertTokenType } from './common';
