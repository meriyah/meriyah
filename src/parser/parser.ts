import { convertTokenType } from '../lexer';
import { Token } from '../token';
import type * as ESTree from '../estree';
import {
  Flags,
  type OnComment,
  type OnInsertedSemicolon,
  type OnToken,
  type Location,
  type AssignmentKind,
  type DestructuringKind,
} from '../common';

export type ParserOptions = {
  shouldAddLoc?: boolean;
  shouldAddRanges?: boolean;
  /**
   * Used together with source maps. File containing the code being parsed
   */
  sourceFile?: string;
  /**
   * Holds either a function or array used on every comment
   */
  onComment?: OnComment;
  /**
   * Function invoked with the character offset when automatic semicolon insertion occurs
   */
  onInsertedSemicolon?: OnInsertedSemicolon;
  /**
   * Holds either a function or array used on every token
   */
  onToken?: OnToken;
};

export class Parser {
  private lastOnToken: [string, number, number, ESTree.SourceLocation] | null = null;

  token = Token.EOF;
  /**
   * The mutable parser flags, in case any flags need passed by reference.
   */
  flags = Flags.None;
  /**
   * The current index
   */
  index = 0;
  /**
   * Beginning of current line
   */
  line = 1;

  /**
   * Beginning of current column
   */
  column = 0;

  /**
   * Start position of whitespace/comment before current token
   */
  startIndex = 0;

  /**
   * The end of the source code
   */
  end = 0;

  /**
   * Start position of text of current token
   */
  tokenIndex = 0;

  /**
   * Start position of the column before newline
   */
  startColumn = 0;

  /**
   * Position in the input code of the first character after the last newline
   */
  tokenColumn = 0;

  /**
   * The number of newlines
   */
  tokenLine = 1;

  /**
   * Start position of text of current token
   */
  startLine = 1;

  /**
   * Holds the scanned token value
   */
  tokenValue: any = '';

  /**
   * Holds the raw text that have been scanned by the lexer
   */
  tokenRaw = '';

  /**
   * Holds the regExp info text that have been collected by the lexer
   */
  tokenRegExp: void | {
    pattern: string;
    flags: string;
  } = void 0;

  /**
   * The code point at the current index
   */
  currentChar = 0;

  /**
   *  https://tc39.es/ecma262/#sec-module-semantics-static-semantics-exportednames
   */
  exportedNames: Record<string, number> = {};

  /**
   * https://tc39.es/ecma262/#sec-exports-static-semantics-exportedbindings
   */

  exportedBindings: Record<string, number> = {};

  /**
   * Assignable state
   */
  assignable: AssignmentKind | DestructuringKind = 1;

  /**
   * Destructuring state
   */
  destructible: AssignmentKind | DestructuringKind = 0;

  /**
   * Holds leading decorators before "export" or "class" keywords
   */
  leadingDecorators: ESTree.Decorator[] = [];

  constructor(
    /**
     * The source code to be parsed
     */
    public readonly source: string,
    public readonly options: ParserOptions = {},
  ) {
    this.end = source.length;
    this.currentChar = source.charCodeAt(0);
  }

  /**
   * Get the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  getToken() {
    return this.token;
  }

  /**
   * Set the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  setToken(value: Token, replaceLast = false) {
    this.token = value;

    const { onToken } = this.options;

    if (onToken) {
      if (value !== Token.EOF) {
        const loc = {
          start: {
            line: this.tokenLine,
            column: this.tokenColumn,
          },
          end: {
            line: this.line,
            column: this.column,
          },
        };

        if (!replaceLast && this.lastOnToken) {
          onToken(...this.lastOnToken);
        }
        this.lastOnToken = [convertTokenType(value), this.tokenIndex, this.index, loc];
      } else {
        if (this.lastOnToken) {
          onToken(...this.lastOnToken);
          this.lastOnToken = null;
        }
      }
    }
    return value;
  }

  get tokenStart(): Location {
    return {
      index: this.tokenIndex,
      line: this.tokenLine,
      column: this.tokenColumn,
    };
  }

  finishNode<T extends ESTree.Node>(node: T, start: Location, end: Location | void): T {
    if (this.options.shouldAddRanges) {
      node.start = start.index;
      const endIndex = end ? end.index : this.startIndex;
      node.end = endIndex;
      node.range = [start.index, endIndex];
    }

    if (this.options.shouldAddLoc) {
      node.loc = {
        start: {
          line: start.line,
          column: start.column,
        },
        end: end ? { line: end.line, column: end.column } : { line: this.startLine, column: this.startColumn },
      };

      if (this.options.sourceFile) {
        node.loc.source = this.options.sourceFile;
      }
    }

    return node;
  }
}

export function pushComment(comments: ESTree.Comment[], options: ParserOptions): OnComment {
  return function (type: ESTree.CommentType, value: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const comment: ESTree.Comment = {
      type,
      value,
    };

    if (options.shouldAddRanges) {
      comment.start = start;
      comment.end = end;
      comment.range = [start, end];
    }
    if (options.shouldAddLoc) {
      comment.loc = loc;
    }
    comments.push(comment);
  };
}

export function pushToken(tokens: any[], options: ParserOptions): OnToken {
  return function (type: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const token: any = {
      token: type,
    };

    if (options.shouldAddRanges) {
      token.start = start;
      token.end = end;
      token.range = [start, end];
    }
    if (options.shouldAddLoc) {
      token.loc = loc;
    }
    tokens.push(token);
  };
}
