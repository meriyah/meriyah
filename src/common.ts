import { Errors } from './errors';
import { nextToken } from './lexer/scan';
import { type Parser } from './parser/parser';
import { KeywordDescTable, Token } from './token';

/**
 * The core context, passed around everywhere as a simple immutable bit set
 */
export const enum Context {
  None = 0,
  Strict = 1 << 0,
  Module = 1 << 1, // Current code should be parsed as a module body
  InSwitch = 1 << 2,
  InGlobal = 1 << 3,
  InClass = 1 << 4,
  AllowRegExp = 1 << 5,
  TaggedTemplate = 1 << 6,
  InIteration = 1 << 7,
  SuperProperty = 1 << 8,
  SuperCall = 1 << 9,
  InYieldContext = 1 << 10,
  InAwaitContext = 1 << 11,
  InReturnContext = 1 << 12,
  InArgumentList = 1 << 13,
  InConstructor = 1 << 14,
  InMethodOrFunction = 1 << 15,
  AllowNewTarget = 1 << 16,
  DisallowIn = 1 << 17,
  AllowEscapedKeyword = 1 << 18,
  InStaticBlock = 1 << 19,
}

/**
 * Masks to track the property kind
 */
export const enum PropertyKind {
  None = 0,
  Method = 1 << 0,
  Computed = 1 << 1,
  Shorthand = 1 << 2,
  Generator = 1 << 3,
  Async = 1 << 4,
  Static = 1 << 5,
  Constructor = 1 << 6,
  ClassField = 1 << 7,
  Getter = 1 << 8,
  Setter = 1 << 9,
  Accessor = 1 << 10,
  Extends = 1 << 11,
  Literal = 1 << 12,
  PrivateField = 1 << 13,
  GetSet = Getter | Setter,
}

/**
 * Masks to track the binding kind
 */
export const enum BindingKind {
  None = 0,
  ArgumentList = 1 << 0,
  Empty = 1 << 1,
  Variable = 1 << 2,
  Let = 1 << 3,
  Const = 1 << 4,
  Class = 1 << 5,
  FunctionLexical = 1 << 6,
  FunctionStatement = 1 << 7,
  CatchPattern = 1 << 8,
  CatchIdentifier = 1 << 9,
  Async = 1 << 10,
  Generator = 1 << 10,
  AsyncFunctionLexical = Async | FunctionLexical,
  GeneratorFunctionLexical = Generator | FunctionLexical,
  AsyncGeneratorFunctionLexical = Async | Generator | FunctionLexical,
  CatchIdentifierOrPattern = CatchIdentifier | CatchPattern,
  LexicalOrFunction = Variable | FunctionLexical,
  LexicalBinding = Let | Const | FunctionLexical | FunctionStatement | Class,
}

/**
 * The masks to track where something begins. E.g. statements, declarations or arrows.
 */
export const enum Origin {
  None = 0,
  Statement = 1 << 0,
  BlockStatement = 1 << 1,
  TopLevel = 1 << 2,
  Declaration = 1 << 3,
  Arrow = 1 << 4,
  ForStatement = 1 << 5,
  Export = 1 << 6,
}

/**
 * Masks to track the assignment kind
 */
export const enum AssignmentKind {
  None = 0,
  Assignable = 1 << 0,
  CannotAssign = 1 << 1,
}

/**
 * Masks to track the destructuring kind
 */
export const enum DestructuringKind {
  None = 0,
  HasToDestruct = 1 << 3,
  // "Cannot" rather than "can" so that this flag can be ORed together across
  // multiple characters.
  CannotDestruct = 1 << 4,
  // Only destructible if assignable
  Assignable = 1 << 5,
  // `__proto__` is a special case and only valid to parse if destructible
  SeenProto = 1 << 6,
  Await = 1 << 7,
  Yield = 1 << 8,
}

/**
 * The mutable parser flags, in case any flags need passed by reference.
 */
export const enum Flags {
  None = 0,
  NewLine = 1 << 0,
  HasConstructor = 1 << 5,
  Octal = 1 << 6,
  NonSimpleParameterList = 1 << 7,
  HasStrictReserved = 1 << 8,
  StrictEvalArguments = 1 << 9,
  DisallowCall = 1 << 10,
  HasOptionalChaining = 1 << 11,
  EightAndNine = 1 << 12,
}

export const enum HoistedClassFlags {
  None,
  Hoisted = 1 << 0,
  Export = 1 << 1,
}

export const enum HoistedFunctionFlags {
  None,
  Hoisted = 1 << 0,
  Export = 1 << 1,
}

/**
 * Check for automatic semicolon insertion according to the rules
 * given in `ECMA-262, section 11.9`.
 *
 * @param parser Parser object
 * @param context Context masks
 */

export function matchOrInsertSemicolon(parser: Parser, context: Context): void {
  if ((parser.flags & Flags.NewLine) === 0 && (parser.getToken() & Token.IsAutoSemicolon) !== Token.IsAutoSemicolon) {
    parser.report(Errors.UnexpectedToken, KeywordDescTable[parser.getToken() & Token.Type]);
  }

  if (!consumeOpt(parser, context, Token.Semicolon)) {
    // Automatic semicolon insertion has occurred
    parser.options.onInsertedSemicolon?.(parser.startIndex);
  }
}

export function isValidStrictMode(parser: Parser, index: number, tokenIndex: number, tokenValue: string): 0 | 1 {
  if (index - tokenIndex < 13 && tokenValue === 'use strict') {
    if ((parser.getToken() & Token.IsAutoSemicolon) === Token.IsAutoSemicolon || parser.flags & Flags.NewLine) {
      return 1;
    }
  }
  return 0;
}

/**
 * Consumes the current token if the current token kind is
 * the specified `kind` and returns `0`. Otherwise returns `1`.
 *
 * @param parser Parser state
 * @param context Context masks
 * @param token The type of token to consume
 */
export function optionalBit(parser: Parser, context: Context, t: Token): 0 | 1 {
  if (parser.getToken() !== t) return 0;
  nextToken(parser, context);
  return 1;
}

/** Consumes the current token if the current token kind is
 * the specified `kind` and returns `true`. Otherwise returns
 * `false`.
 *
 * @param parser Parser state
 * @param context Context masks
 * @param token The type of token to consume
 */
export function consumeOpt(parser: Parser, context: Context, t: Token): boolean {
  if (parser.getToken() !== t) return false;
  nextToken(parser, context);
  return true;
}

/**
 * Consumes the current token. If the current token kind is not
 * the specified `kind`, an error will be reported.
 *
 * @param parser Parser state
 * @param context Context masks
 * @param t The type of token to consume
 */
export function consume(parser: Parser, context: Context, t: Token): void {
  if (parser.getToken() !== t) parser.report(Errors.ExpectedToken, KeywordDescTable[t & Token.Type]);
  nextToken(parser, context);
}

/**
 * Transforms a `LeftHandSideExpression` into a `AssignmentPattern` if possible,
 * otherwise it returns the original tree.
 *
 * @param parser Parser state
 * @param {*} node
 */
export function reinterpretToPattern(parser: Parser, node: any): void {
  switch (node.type) {
    case 'ArrayExpression': {
      node.type = 'ArrayPattern';
      const { elements } = node;
      for (let i = 0, n = elements.length; i < n; ++i) {
        const element = elements[i];
        if (element) reinterpretToPattern(parser, element);
      }
      return;
    }
    case 'ObjectExpression': {
      node.type = 'ObjectPattern';
      const { properties } = node;
      for (let i = 0, n = properties.length; i < n; ++i) {
        reinterpretToPattern(parser, properties[i]);
      }
      return;
    }
    case 'AssignmentExpression':
      node.type = 'AssignmentPattern';
      if (node.operator !== '=') parser.report(Errors.InvalidDestructuringTarget);
      delete node.operator;
      reinterpretToPattern(parser, node.left);
      return;
    case 'Property':
      reinterpretToPattern(parser, node.value);
      return;
    case 'SpreadElement':
      node.type = 'RestElement';
      reinterpretToPattern(parser, node.argument);
    // No default
  }
}

/**
 * Validates binding identifier
 *
 * @param parser Parser state
 * @param context Context masks
 * @param type Binding type
 * @param token Token
 */

export function validateBindingIdentifier(
  parser: Parser,
  context: Context,
  kind: BindingKind,
  t: Token,
  skipEvalArgCheck: 0 | 1,
): void {
  if (context & Context.Strict) {
    if ((t & Token.FutureReserved) === Token.FutureReserved) {
      parser.report(Errors.UnexpectedStrictReserved);
    }

    if (!skipEvalArgCheck && (t & Token.IsEvalOrArguments) === Token.IsEvalOrArguments) {
      parser.report(Errors.StrictEvalArguments);
    }
  }

  if ((t & Token.Reserved) === Token.Reserved || t === Token.EscapedReserved) {
    parser.report(Errors.KeywordNotId);
  }

  // The BoundNames of LexicalDeclaration and ForDeclaration must not
  // contain 'let'. (CatchParameter is the only lexical binding form
  // without this restriction.)
  if (kind & (BindingKind.Let | BindingKind.Const) && (t & Token.Type) === (Token.LetKeyword & Token.Type)) {
    parser.report(Errors.InvalidLetConstBinding);
  }

  if (context & (Context.InAwaitContext | Context.Module) && t === Token.AwaitKeyword) {
    parser.report(Errors.AwaitIdentInModuleOrAsyncFunc);
  }

  if (context & (Context.InYieldContext | Context.Strict) && t === Token.YieldKeyword) {
    parser.report(Errors.DisallowedInContext, 'yield');
  }
}

export function validateFunctionName(parser: Parser, context: Context, t: Token): void {
  if (context & Context.Strict) {
    if ((t & Token.FutureReserved) === Token.FutureReserved) {
      parser.report(Errors.UnexpectedStrictReserved);
    }

    if ((t & Token.IsEvalOrArguments) === Token.IsEvalOrArguments) {
      parser.report(Errors.StrictEvalArguments);
    }

    if (t === Token.EscapedFutureReserved) {
      parser.report(Errors.InvalidEscapedKeyword);
    }

    if (t === Token.EscapedReserved) {
      parser.report(Errors.InvalidEscapedKeyword);
    }
  }

  if ((t & Token.Reserved) === Token.Reserved) {
    parser.report(Errors.KeywordNotId);
  }

  if (context & (Context.InAwaitContext | Context.Module) && t === Token.AwaitKeyword) {
    parser.report(Errors.AwaitIdentInModuleOrAsyncFunc);
  }

  if (context & (Context.InYieldContext | Context.Strict) && t === Token.YieldKeyword) {
    parser.report(Errors.DisallowedInContext, 'yield');
  }
}

/**
 * Validates binding identifier
 *
 * @param parser Parser state
 * @param context Context masks
 * @param t Token
 */

export function isStrictReservedWord(parser: Parser, context: Context, t: Token): boolean {
  if (t === Token.AwaitKeyword) {
    if (context & (Context.InAwaitContext | Context.Module)) parser.report(Errors.AwaitIdentInModuleOrAsyncFunc);
    parser.destructible |= DestructuringKind.Await;
  }

  if (t === Token.YieldKeyword && context & Context.InYieldContext) parser.report(Errors.DisallowedInContext, 'yield');

  return (
    (t & Token.Reserved) === Token.Reserved ||
    (t & Token.FutureReserved) === Token.FutureReserved ||
    t == Token.EscapedFutureReserved
  );
}

/**
 * Checks if the property has any private field key
 *
 * @param parser Parser object
 * @param context  Context masks
 */
export function isPropertyWithPrivateFieldKey(expr: any): boolean {
  return !expr.property ? false : expr.property.type === 'PrivateIdentifier';
}

/**
 * Checks if a label in `LabelledStatement` are valid or not
 *
 * @param parser Parser state
 * @param labels Object holding the labels
 * @param name Current label
 * @param isIterationStatement
 */
export function isValidLabel(parser: Parser, labels: any, name: string, isIterationStatement: 0 | 1): 0 | 1 {
  while (labels) {
    if (labels['$' + name]) {
      if (isIterationStatement) parser.report(Errors.InvalidNestedStatement);
      return 1;
    }
    if (isIterationStatement && labels.loop) isIterationStatement = 0;
    labels = labels['$'];
  }

  return 0;
}

/**
 * Checks if current label already have been declared, and if not
 * declare it
 *
 * @param parser Parser state
 * @param labels Object holding the labels
 * @param name Current label
 */
export function validateAndDeclareLabel(parser: Parser, labels: any, name: string): void {
  let set = labels;
  while (set) {
    if (set['$' + name]) parser.report(Errors.LabelRedeclaration, name);
    set = set['$'];
  }

  labels['$' + name] = 1;
}

/** @internal */
export function isEqualTagName(elementName: any): any {
  switch (elementName.type) {
    case 'JSXIdentifier':
      return elementName.name;
    case 'JSXNamespacedName':
      return elementName.namespace + ':' + elementName.name;
    case 'JSXMemberExpression':
      return isEqualTagName(elementName.object) + '.' + isEqualTagName(elementName.property);
    /* istanbul ignore next */
    default:
    // ignore
  }
}

export function isValidIdentifier(context: Context, t: Token): boolean {
  if (context & (Context.Strict | Context.InYieldContext)) {
    // Module code is also "strict mode code"
    if (context & Context.Module && t === Token.AwaitKeyword) return false;
    if (context & Context.InYieldContext && t === Token.YieldKeyword) return false;
    return (t & Token.Contextual) === Token.Contextual;
  }

  return (t & Token.Contextual) === Token.Contextual || (t & Token.FutureReserved) === Token.FutureReserved;
}

export function classifyIdentifier(parser: Parser, context: Context, t: Token): any {
  if ((t & Token.IsEvalOrArguments) === Token.IsEvalOrArguments) {
    if (context & Context.Strict) parser.report(Errors.StrictEvalArguments);
    parser.flags |= Flags.StrictEvalArguments;
  }

  if (!isValidIdentifier(context, t)) parser.report(Errors.Unexpected);
}

export type Location = {
  readonly index: number;
  readonly line: number;
  readonly column: number;
};

export function getOwnProperty<T>(object: Record<string | number, T>, key: string | number): T | undefined {
  return Object.hasOwn(object, key) ? object[key] : undefined;
}
