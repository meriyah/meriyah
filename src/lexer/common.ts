import { Chars } from '../chars';
import { Flags } from '../common';
import { type Parser } from '../parser/parser';
import { Token } from '../token';

export const enum LexerState {
  None = 0,
  NewLine = 1 << 0,
  LastIsCR = 1 << 2,
}

export const enum NumberKind {
  ImplicitOctal = 1 << 0,
  Binary = 1 << 1,
  Octal = 1 << 2,
  Hex = 1 << 3,
  Decimal = 1 << 4,
  NonOctalDecimal = 1 << 5,
  Float = 1 << 6,
  ValidBigIntKind = 1 << 7,
  DecimalNumberKind = Decimal | NonOctalDecimal,
}

/**
 * Advances this lexer's current index.
 *
 * @param parser The parser instance
 */
export function advanceChar(parser: Parser): number {
  parser.column++;
  return (parser.currentChar = parser.source.charCodeAt(++parser.index));
}

/**
 * Consumes possible surrogate pair, return the merged code point or 0
 *
 * @param parser The parser instance
 */
export function consumePossibleSurrogatePair(parser: Parser): number {
  // See: https://tc39.github.io/ecma262/#sec-ecmascript-language-types-string-type
  const hi = parser.currentChar;
  // Not surrogate
  if ((hi & 0xfc00) !== Chars.LeadSurrogateMin) return Chars.Null;
  const lo = parser.source.charCodeAt(parser.index + 1);
  // Unpaired surrogate
  if ((lo & 0xfc00) !== Chars.TrailSurrogateMin) return Chars.Null;
  // Merged code point
  return Chars.NonBMPMin + ((hi & 0x3ff) << 10) + (lo & 0x3ff);
}

/**
 * Use to consume a line feed instead of `scanNewLine`.
 */
export function consumeLineFeed(parser: Parser, state: LexerState): void {
  parser.currentChar = parser.source.charCodeAt(++parser.index);
  parser.flags |= Flags.NewLine;
  if ((state & LexerState.LastIsCR) === 0) {
    parser.column = 0;
    parser.line++;
  }
}

export function scanNewLine(parser: Parser): void {
  parser.flags |= Flags.NewLine;
  parser.currentChar = parser.source.charCodeAt(++parser.index);
  parser.column = 0;
  parser.line++;
}

// ECMA-262 11.2 White Space
// gC=Zs, U+0009, U+000B, U+000C, U+FEFF
export function isExoticECMAScriptWhitespace(ch: number): boolean {
  return (
    ch === Chars.NonBreakingSpace ||
    ch === Chars.ZeroWidthNoBreakSpace ||
    ch === Chars.NextLine ||
    ch === Chars.Ogham ||
    (ch >= Chars.EnQuad && ch <= Chars.ZeroWidthSpace) ||
    ch === Chars.NarrowNoBreakSpace ||
    ch === Chars.MathematicalSpace ||
    ch === Chars.IdeographicSpace ||
    ch === Chars.ThinSpace ||
    ch === Chars.ByteOrderMark
  );
}

/**
 * Converts a value to a hex value
 *
 * @param code CodePoint
 */
export function toHex(code: number): number {
  return code < Chars.UpperA ? code - Chars.Zero : (code - Chars.UpperA + 10) & 0xf;
}

/**
 * Converts a token to a string representation
 *
 * @param t Token
 */
export function convertTokenType(t: Token): string {
  switch (t) {
    case Token.NumericLiteral:
      return 'NumericLiteral';
    case Token.StringLiteral:
      return 'StringLiteral';
    case Token.FalseKeyword:
    case Token.TrueKeyword:
      return 'BooleanLiteral';
    case Token.NullKeyword:
      return 'NullLiteral';
    case Token.RegularExpression:
      return 'RegularExpression';
    case Token.TemplateContinuation:
    case Token.TemplateSpan:
    case Token.Template:
      return 'TemplateLiteral';
    default:
      if ((t & Token.IsIdentifier) === Token.IsIdentifier) return 'Identifier';
      if ((t & Token.Keyword) === Token.Keyword) return 'Keyword';

      return 'Punctuator';
  }
}
