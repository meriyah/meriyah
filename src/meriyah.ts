import { Context } from './common';
import { parseSource, Options } from './parser';
import * as ESTree from './estree';
// Current version
import { version } from '../package.json';

/**
 * Parse a script, optionally with various options.
 */
export function parseScript(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

/**
 * Parse a module, optionally with various options.
 */
export function parseModule(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.Strict | Context.Module);
}

/**
 * Parse a module or a script, optionally with various options.
 */
export function parse(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

export { Options, ESTree, version };
