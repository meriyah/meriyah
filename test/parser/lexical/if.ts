import { Context } from '../../../src/common';
import { pass, fail } from '../../test-utils';
import * as t from 'assert';
import { parseSource } from '../../../src/parser';

describe('Lexical - If', () => {
  fail('Lexical - If (fail)', [
    ['if (x) var foo = 1; let foo = 1;', Context.OptionsLexical],
    ['if (x) {} else var foo = 1; let foo = 1;', Context.OptionsLexical],
    ['if (x) var foo = 1; else {} let foo = 1;', Context.OptionsLexical],
    ['if (x) var foo = 1; let foo = 1;', Context.OptionsLexical],

    ['if (x) {} else if (y) {} else var foo = 1; let foo = 1;', Context.OptionsLexical],
    ['if (x) { if (y) var foo = 1; } let foo = 1;', Context.OptionsLexical],
    ['const x = a; function x(){};', Context.OptionsLexical],
    ['if (x) var foo = 1; let foo = 1;', Context.OptionsWebCompat | Context.OptionsLexical],
    ['if (x) var foo = 1; let foo = 1;', Context.OptionsWebCompat | Context.OptionsLexical],
    ['if (x) {} else var foo = 1; let foo = 1;', Context.OptionsWebCompat | Context.OptionsLexical],
    ['if (x) var foo = 1; else {} let foo = 1;', Context.OptionsWebCompat | Context.OptionsLexical],
    ['if (x) {} else if (y) {} else var foo = 1; let foo = 1;', Context.OptionsWebCompat | Context.OptionsLexical],
    ['if (x) { if (y) var foo = 1; } let foo = 1;', Context.OptionsWebCompat | Context.OptionsLexical]
  ]);

  for (const arg of ['if (x) var foo = 1; var foo = 1;']) {
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, undefined, Context.OptionsWebCompat | Context.OptionsLexical);
      });
    });
  }
});
