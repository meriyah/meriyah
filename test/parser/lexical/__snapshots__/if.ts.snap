// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Lexical - If > Lexical - If (fail) > const x = a; function x(){}; 1`] = `
"SyntaxError [1:22-1:23]: Duplicate binding 'x'
> 1 | const x = a; function x(){};
    |                       ^ Duplicate binding 'x'"
`;

exports[`Lexical - If > Lexical - If (fail) > const x = a; function x(){}; 2`] = `
"SyntaxError [1:22-1:23]: Duplicate binding 'x'
> 1 | const x = a; function x(){};
    |                       ^ Duplicate binding 'x'"
`;

exports[`Lexical - If > Lexical - If (fail) > do async function f(){} while (x); 1`] = `
"SyntaxError [1:9-1:17]: Async functions can only be declared at the top level or inside a block
> 1 | do async function f(){} while (x);
    |          ^^^^^^^^ Async functions can only be declared at the top level or inside a block"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) { if (y) var foo = 1; } let foo = 1; 1`] = `
"SyntaxError [1:39-1:40]: Duplicate binding 'foo'
> 1 | if (x) { if (y) var foo = 1; } let foo = 1;
    |                                        ^ Duplicate binding 'foo'"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) { if (y) var foo = 1; } let foo = 1; 2`] = `
"SyntaxError [1:39-1:40]: Duplicate binding 'foo'
> 1 | if (x) { if (y) var foo = 1; } let foo = 1;
    |                                        ^ Duplicate binding 'foo'"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) { if (y) var foo = 1; } let foo = 1; 3`] = `
"SyntaxError [1:39-1:40]: Duplicate binding 'foo'
> 1 | if (x) { if (y) var foo = 1; } let foo = 1;
    |                                        ^ Duplicate binding 'foo'"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) { if (y) var foo = 1; } let foo = 1; 4`] = `
"SyntaxError [1:39-1:40]: Duplicate binding 'foo'
> 1 | if (x) { if (y) var foo = 1; } let foo = 1;
    |                                        ^ Duplicate binding 'foo'"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) {} else if (y) {} else var foo = 1; let foo = 1; 1`] = `
"SyntaxError [1:51-1:52]: Duplicate binding 'foo'
> 1 | if (x) {} else if (y) {} else var foo = 1; let foo = 1;
    |                                                    ^ Duplicate binding 'foo'"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) {} else if (y) {} else var foo = 1; let foo = 1; 2`] = `
"SyntaxError [1:51-1:52]: Duplicate binding 'foo'
> 1 | if (x) {} else if (y) {} else var foo = 1; let foo = 1;
    |                                                    ^ Duplicate binding 'foo'"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) {} else var foo = 1; let foo = 1; 1`] = `
"SyntaxError [1:36-1:37]: Duplicate binding 'foo'
> 1 | if (x) {} else var foo = 1; let foo = 1;
    |                                     ^ Duplicate binding 'foo'"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) {} else var foo = 1; let foo = 1; 2`] = `
"SyntaxError [1:36-1:37]: Duplicate binding 'foo'
> 1 | if (x) {} else var foo = 1; let foo = 1;
    |                                     ^ Duplicate binding 'foo'"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) async function f(){} 1`] = `
"SyntaxError [1:13-1:21]: Async functions can only be declared at the top level or inside a block
> 1 | if (x) async function f(){}
    |              ^^^^^^^^ Async functions can only be declared at the top level or inside a block"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) var foo = 1; else {} let foo = 1; 1`] = `
"SyntaxError [1:36-1:37]: Duplicate binding 'foo'
> 1 | if (x) var foo = 1; else {} let foo = 1;
    |                                     ^ Duplicate binding 'foo'"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) var foo = 1; else {} let foo = 1; 2`] = `
"SyntaxError [1:36-1:37]: Duplicate binding 'foo'
> 1 | if (x) var foo = 1; else {} let foo = 1;
    |                                     ^ Duplicate binding 'foo'"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) var foo = 1; let foo = 1; 1`] = `
"SyntaxError [1:28-1:29]: Duplicate binding 'foo'
> 1 | if (x) var foo = 1; let foo = 1;
    |                             ^ Duplicate binding 'foo'"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) var foo = 1; let foo = 1; 2`] = `
"SyntaxError [1:28-1:29]: Duplicate binding 'foo'
> 1 | if (x) var foo = 1; let foo = 1;
    |                             ^ Duplicate binding 'foo'"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) var foo = 1; let foo = 1; 3`] = `
"SyntaxError [1:28-1:29]: Duplicate binding 'foo'
> 1 | if (x) var foo = 1; let foo = 1;
    |                             ^ Duplicate binding 'foo'"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) var foo = 1; let foo = 1; 4`] = `
"SyntaxError [1:28-1:29]: Duplicate binding 'foo'
> 1 | if (x) var foo = 1; let foo = 1;
    |                             ^ Duplicate binding 'foo'"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) var foo = 1; let foo = 1; 5`] = `
"SyntaxError [1:28-1:29]: Duplicate binding 'foo'
> 1 | if (x) var foo = 1; let foo = 1;
    |                             ^ Duplicate binding 'foo'"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) x;
else async function *f(){} 1`] = `
"SyntaxError [2:11-2:19]: Async functions can only be declared at the top level or inside a block
  1 | if (x) x;
> 2 | else async function *f(){}
    |            ^^^^^^^^ Async functions can only be declared at the top level or inside a block"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) x;
else async function f(){} 1`] = `
"SyntaxError [2:11-2:19]: Async functions can only be declared at the top level or inside a block
  1 | if (x) x;
> 2 | else async function f(){}
    |            ^^^^^^^^ Async functions can only be declared at the top level or inside a block"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) x;
else function *f(){} 1`] = `
"SyntaxError [2:5-2:13]: Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement
  1 | if (x) x;
> 2 | else function *f(){}
    |      ^^^^^^^^ Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement"
`;

exports[`Lexical - If > Lexical - If (fail) > if (x) x;
else function(){} 1`] = `
"SyntaxError [2:5-2:13]: Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement
  1 | if (x) x;
> 2 | else function(){}
    |      ^^^^^^^^ Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement"
`;
